misc/python/materialize/checks/range.py

```diff
@@ -0,0 +1,110 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from textwrap import dedent
+from typing import List
+
+from materialize.checks.actions import Testdrive
+from materialize.checks.checks import Check
+
+
+class Range(Check):
```

Comment:
This check is currently only about `INT8RANGE`. Other ranges exist too.

I intend to generalize this class and create subclasses for the different types.

Reply:
Yes, in an ideal world one would subclass or otherwise organize things so that they are not jammed together. Unfortunartely, in Mz, maintaining a materialized view (and source, even more) is a resource-consuming operation (Mz will perform various housekeeping tasks involving multiple round-trips to CRDB every second for every materialized view), so it would not be possible to devote one materialized view per type. So there is no option but to test all types together within the same queries -- see how `numeric_types.py` does it for all integer types :-(

Reply:
Not sure if I fully understood it.

Are you suggesting that test execution will become to expensive if I add multiple test classes with their own materialized view? Aren't the materialized views torn down at the end of the test?

Ok, will have a look at `numeric_types.py`.

Reply:
It would be fine to land it with int8range at first. Going through all of the combinations can be expensive, and this is being run in many instances. One approach I used to get more coverage without using more test time in identifiers.py was to randomly pick one set: https://github.com/MaterializeInc/materialize/blob/8a2d1ad817649fd817ddad391ce17bfaca120f3a/misc/python/materialize/checks/identifiers.py#L210

Reply:
The problem is that, in CI, all the Checks are performed together against the same Mz instance, and all of their views exist throughout the test and all the way until the end of the test. Those are all those views:

```
cd misc/python/materialize/checks
grep -i 'MATERIALIZED VIEW' * | grep -v SHOW | wc
grep: __pycache__: Is a directory
    122    1850   17531
```

Mz performs a bunch of operations per second for every existing view and this adds up to a lot of CPU and CRDB usage. If you just run the platform checks framework all the way to the end, while keeping the Mz instance running after the end of the test:

```
diff --git a/test/platform-checks/mzcompose.py b/test/platform-checks/mzcompose.py
index 2653906f2..4ce393c25 100644
--- a/test/platform-checks/mzcompose.py
+++ b/test/platform-checks/mzcompose.py
@@ -65,8 +65,9 @@ def setup(c: Composition) -> None:
 
 
 def teardown(c: Composition) -> None:
-    c.rm(*[s.name for s in SERVICES], stop=True, destroy_volumes=True)
-    c.rm_volumes("mzdata", "tmp", force=True)
+     pass
+#    c.rm(*[s.name for s in SERVICES], stop=True, destroy_volumes=True)
+#    c.rm_volumes("mzdata", "tmp", force=True)
 
 
 def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
```

```
cd test/platform-checks
./mzcompose run default --scenario=NoRestartNoUpgrade
```

you can see that CPU and CRDB usage is through the roof even though the database, technically speaking, is idle after the test is complete.

To account for the cost of having many materialized views, Mz by default is limited to I think 25 materialized views and  the Checks framework issues a separate `ALTER CLUSTER` statement to lift that restriction. Restrictions also exist on the number of tables and the number of sources/sinks, as all of those are also consuming resources.

At some point, not sure if this is still the case, if Mz is unable to complete within 1 second all the housekeeping tasks that have to happen within 1 second, the entire system will lock up. And the quantity of those housekeeping tasks is proportional to the number of materialized views/tables/sources/sinks.